### ISSUE TITLE
Download from Cart Results in NoClassDefFoundError

### ISSUE THREAD

**Issue by dsandqui** [adobe/asset-share-commons#834]
**Describe the bug**
When downloading files from the Cart, after selecting the download options and clicking the Download button, the result is a NoClassDefFoundError exception for the ACS AEM Commons class com/adobe/acs/commons/util/BufferedServletOutput:

```
2022-10-26 11:45:18.905 DEBUG [com.adobe.aem.commons.assetshare.components.actions.download.impl.DownloadImpl] Initing DownloadImpl withe RequireAem configurations: [ distribution = classic ] and [ serviceType = author ]
2022-10-26 11:45:18.908 DEBUG [com.adobe.aem.commons.assetshare.components.actions.download.impl.DownloadImpl] Max allowed content size (in bytes) [ -1 ]
26.10.2022 11:45:26.754 *ERROR* [[0:0:0:0:0:0:0:1] [1666802726727] POST /content/asset-share-commons/en/dark/actions/download/_jcr_content/root/responsivegrid/download.download-asset-renditions.zip HTTP/1.1] org.apache.sling.engine.impl.SlingRequestProcessorImpl service: Uncaught Throwable
java.lang.NoClassDefFoundError: com/adobe/acs/commons/util/BufferedServletOutput
	at com.adobe.acs.commons.util.BufferedSlingHttpServletResponse.<init>(BufferedSlingHttpServletResponse.java:52)
	at com.adobe.aem.commons.assetshare.content.renditions.download.impl.AssetRenditionDownloadResponse.<init>(AssetRenditionDownloadResponse.java:39) [com.adobe.aem.commons.assetshare.core:2.4.0]
	at com.adobe.aem.commons.assetshare.content.renditions.download.impl.AssetRenditionStreamerImpl.sendDispatchForAssetRendition(AssetRenditionStreamerImpl.java:106) [com.adobe.aem.commons.assetshare.core:2.4.0]
	at com.adobe.aem.commons.assetshare.content.renditions.download.impl.AssetRenditionStreamerImpl.getAssetRendition(AssetRenditionStreamerImpl.java:75) [com.adobe.aem.commons.assetshare.core:2.4.0]
	at com.adobe.aem.commons.assetshare.content.renditions.download.impl.AssetRenditionsZipperImpl.execute(AssetRenditionsZipperImpl.java:115) [com.adobe.aem.commons.assetshare.core:2.4.0]
	at com.adobe.aem.commons.assetshare.content.renditions.download.impl.AssetRenditionsDownloadServlet.doPost(AssetRenditionsDownloadServlet.java:112) [com.adobe.aem.commons.assetshare.core:2.4.0]
```

The exception occurs in both the customer solution I am supporting and after deploying the "OOTB" assetshare.ui.content.sample-2.4.0.zip.  It also occurs on both my local AEM instance and the AEMaaCS dev environment.  I've tested this on the author instance in both environments, but have not yet tested on a publish instance.

I also tried installing the ACS AEM Commons bundle and even added it as a dependency for my core module, but that didn't help. 
I believe that the BufferedServletOutput class is the only ACS AEM Commons class used by the Assetshare bundle v2.4.0.  The exception would indicate that the class is not loaded when invoked, and given that it is not an OSGi service, it may not get loaded automatically.  However, the issue may have something to do related the extraction of the cloud-specific classes into the supporting Fragment bundle, which includes the download functionality.

Given that this is a blocker for my customer, my short term solution may be to create a temporary patch to the Assetshare cloud (fragment) bundle by adding a copy of the BufferedServletOutput class so that it gets loaded by the main bundle's class loader.

**Environment**
* AEM Version: AEMaaCS (latest)
* Asset Share Commons Version: 2.4.0
* Author, Publish or both? Author, but probably both

**Screenshots**
Here is the result of clicking the download button (in the download options modal):

![image](https://user-images.githubusercontent.com/6818091/198094247-d6aca0ef-ec35-4df9-a2f9-3cc9c47e0d55.png)

**Comment by dsandqui** [comment]
I have since found the cause of this issue and at least partial fix.  In the configuration for the bnd-maven-plugin of the asset-share-commons core module pom, there is a setting -includeresource that specifies classes from the acs-aem-commons-bundle that should be embedded instead of embedding the entire bundle.  There are three classes listed in the pom for version 2.4.0.  I added the BufferedServletOutput class to the list and rebuilt the bundle.

After installing it and testing, I got the same exception, except that it was for a different class.  After repeating this process three more times (four classes were added to the -includeresource setting), the download worked correctly.  The classes that I added were:

- BufferedServletOutput
- BufferedServletOutput$ResponseWriteMethod
- ServletOutputStreamWrapper
- BufferedServletResponse

The plugin definition now looks like this:

```
<plugin>
                <groupId>biz.aQute.bnd</groupId>
                <artifactId>bnd-maven-plugin</artifactId>
                <configuration>
                    <bnd><![CDATA[
Bundle-Name: Asset Share Commons - Core
Bundle-Description: Asset Share Commons Core OSGi bundle containing AEM server-side logic
Bundle-SymbolicName: com.adobe.aem.commons.assetshare.core
# conditionalpackage is too coarse, as it will drag in all transitive deps of arbitrary classes in one package, only embed specific classes
-includeresource: @acs-aem-commons-bundle-[0-9.]*.jar!/com/adobe/acs/commons/util/(BufferedSlingHttpServletResponse|BufferedServletOutput|BufferedServletOutput$ResponseWriteMethod|ServletOutputStreamWrapper|BufferedServletResponse|PathInfoUtil|ParameterUtil).*,\
                            ]]></bnd>
                </configuration>
            </plugin>
```

This is a partial fix, as I have only tested locally and only the download from card feature.  There may be additional classes in other features that need to be specified in the configuration.  In addition, I haven't yet tested in the cloud.

Regarding my temporary workaround, I can rebuild the bundle with the new dependent classes embedded without worrying about the baseline plugin complaining.  My only concern is how to install my ASC bundle without the embedded release version of the ASC bundle overwriting it whenever a build is deployed.

**Comment by davidjgonzalez** [comment]
@kwin i think this was changed in the split of the core bundle, any thoughts?

**Comment by kwin** [comment]
This is a regression of https://github.com/adobe/asset-share-commons/commit/013972794fa4be9fbdae3884821f66f4dc5d9fb5#diff-8d04401f1cc51365fe3e32f019cd720135ba920a1a7da7f19e9c9208478701fcR43 which limited the embedded classes.

**Comment by kwin** [comment]
I fail to reproduce. @dsandqui can you please try to describe exactly how to reproduce with https://github.com/adobe/asset-share-commons/tree/develop/ui.content.sample?

**Comment by dsandqui** [comment]
I have been testing on my local AEM author instance, which was built using the latest cloud SDK (2022.10.xxxxxxx).  I’ve also tested on the customer’s AEMaaCS Dev author instance, with the same result.  Here are the steps:

1. Logon to AEM and navigate to Sites.
2. Select asset-share-commons, then en, then dark.
3. Click the checkbox for the dark theme and select edit. You should now see the page at http://localhost:4502/editor.html/content/asset-share-commons/en/dark.html.
4. Switch to Preview mode if it is not already selected.
5. Assuming you have assets deployed to the default dam location, which should be visible in the search results panel, click on the “Add to Cart” link under several of the assets.  Those assets should now be listed in your cart.
6. Open the cart and click the Download Cart button.  The Download Options modal should be displayed.
7. Click on the Original (renditions) checkbox to select it.
8. Click the Download button.
9. You should be redirected to another page / browser tab that displays an error message.
10. Open the error.log file and you should see the ClassDefNotFound exception that was thrown.

**Comment by kwin** [comment]
@dsandqui For me it says "Download queued" after step 8. I was testing with AEM Version 2022.7.8085.20220725T140323Z.

**Comment by davidjgonzalez** [comment]
hi @dsandqui  - @kwin's fix is in 2.4.4 - i checked and it works on 6.5.13 - can you check it out and let us know if it fixes your issue?
